ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -47,6 +48,359 @@ struct PendingRenameOperation
+        }
+    };
+public:
+    KextLogTracer(kauth_action_t vnodeAction, vnode_t vnode) :
+        traceBufferPosition(0),
+        dynamicTraceBuffer(nullptr),
+        dynamicTraceBufferSize(0),
+        discarded(false),
+        willEmitTrace(false)
+    {
+        this->embeddedTraceBuffer[0] = '\0';
+        
+        kauth_action_t mask = atomic_load(&KextLogTracer::traceVnodeActionFilterMask);
+        if (0 == (mask & vnodeAction))
+        {
+            this->discarded = true;
+        }
+        else
+        {
+            char vnodePath[PATH_MAX] = "";
+            if (!ShouldTraceEventsForVnode(vnode, vnodePath))
+            {
+                this->discarded = true;
+            }
+            else
+            {
+                pid_t pid = proc_selfpid();
+                char processName[MAXCOMLEN + 1] = "";
+                proc_selfname(processName, sizeof(processName));
+                
+                if (vnode_isdir(vnode))
+                {
+                    this->Printf(
+                        "Directory vnode '%s' event by process '%s' (PID = %d) action" KextLog_DirectoryVnodeActionFormat,
+                        vnodePath,
+                        processName,
+                        pid,
+                        KextLog_DirectoryVnodeActionArgs(vnodeAction, " "));
+                }
+                else
+                {
+                    this->Printf(
+                        "File vnode '%s' event by process '%s' (PID = %d) action" KextLog_FileVnodeActionFormat,
+                        vnodePath,
+                        processName,
+                        pid,
+                        KextLog_FileVnodeActionArgs(vnodeAction, " "));
+                }
+
+                if (this->traceAllVnodeEvents)
+                {
+                    this->willEmitTrace = true;
+                }
+            }
+        }
+    }
+    
+    void SendProviderMessage(MessageType providerMessage)
+    {
+        if (!this->discarded)
+        {
+            if (!this->willEmitTrace && atomic_load(&KextLogTracer::traceProviderMessagingEvents))
+            {
+                this->willEmitTrace = true;
+            }
+            this->Printf("\nMessage to provider: %u (%s)", providerMessage, MessageTypeString(providerMessage));
+            
+            if (this->willEmitTrace)
+            {
+                this->Emit();
+            }
+        }
+    }
+
+    void SendProviderMessageResult(bool success)
+    {
+        if (!this->discarded)
+        {
+            this->Printf(" -> result: %s", success ? "success" : "failed");
+        }
+    }
+
+    void SetDeniedForCrawler()
+    {
+        if (!this->willEmitTrace && !this->discarded)
+        {
+            if (!atomic_load(&KextLogTracer::traceAllVnodeEvents) && atomic_load(&KextLogTracer::traceDeniedVnodeEvents))
+            {
+                // When tracing only denied events, throw out any denied crawlers as they tend to spam the trace
```

Comment:
This seems ok, but is very good to document as you've done

Reply:
Yeah; we could add an extra option so that those denied events are also logged, but I suspect it's probably very rare that you need this.